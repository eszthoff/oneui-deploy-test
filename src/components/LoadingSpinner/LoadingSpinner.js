import React from 'react';
import PropTypes from 'prop-types';
import bem from '../../packages/bem';
import styles from './LoadingSpinner.scss';
import { CONTEXTS } from '../../constants';

const { block, elem } = bem({
    name: 'LoadingSpinner',
    classnames: styles,
    propsToMods: ['context', 'hidden']
});

const LoadingSpinner = props => {
    const { context, hidden, size, ...rest } = props;
    return (
        <svg
            viewBox={[0, 0, 44, 44]}
            style={(s => {
                if (!s) {
                    return null;
                }
                return {
                    width: s,
                    height: s
                };
            })(size)}
            {...rest}
            {...block(props)}
        >
            <circle cx="22" cy="22" r="20" fill="none" strokeWidth="4" {...elem('path', props)} />
        </svg>
    );
};

LoadingSpinner.propTypes = {
    /** The spinner context (e.g. brand, primary, bad, good etc. - defaults to brand) */
    context: PropTypes.oneOf(CONTEXTS),
    /** Hides the spinner when true */
    hidden: PropTypes.bool,
    /** Custom spinner size (will affect both width and height) */
    size: PropTypes.number
};

LoadingSpinner.defaultProps = {
    context: 'brand',
    hidden: false,
    size: null
};

export default LoadingSpinner;
