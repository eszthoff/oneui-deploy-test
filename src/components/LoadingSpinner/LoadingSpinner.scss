@keyframes rotate {
    100% { transform: rotate(360deg); }
}

@keyframes dash {
    0% {
        stroke-dasharray: 1, 150;
        stroke-dashoffset: 0;
    }
    50% {
        stroke-dasharray: 90, 150;
        stroke-dashoffset: -35;
    }
    100% {
        stroke-dasharray: 90, 150;
        stroke-dashoffset: -124;
    }
}

.LoadingSpinner {
    animation: rotate 2s linear infinite;
    width: 48px;
    height: 48px;
    display: block;

    &__path {
        stroke-dasharray: 1, 150;
        stroke-dashoffset: 0;
        stroke-linecap: round;
        animation: dash 1.5s ease-in-out infinite;
        transition: stroke var(--transition-duration);
    }

    &--context {
        @each $context in (neutral, brand, primary, accent, info, good, warning, bad) {
            &_#{ $context } {
                .LoadingSpinner__path {
                    stroke: var(--color-#{ $context });
                }
            }
        }
    }
}
