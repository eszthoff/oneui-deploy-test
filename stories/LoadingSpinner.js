import React from 'react';
import { storiesOf } from '@storybook/react';
import { boolean, number, select, text, withKnobs } from '@storybook/addon-knobs';
import { LoadingSpinner } from '@textkernel/oneui';
import { CONTEXTS } from '@textkernel/oneui/constants';

storiesOf('LoadingSpinner', module)
    .addDecorator(withKnobs)
    .add('LoadingSpinner', () => (
        <LoadingSpinner
            centerIn={select(
                'Center in...',
                {
                    '': null,
                    'Parent element': 'parent',
                    Viewport: 'viewport'
                },
                null
            )}
            context={select('Context', CONTEXTS, CONTEXTS[1])}
            hidden={boolean('Hidden', false)}
            label={text('Label', 'Loading...')}
            size={number('Size', null)}
        />
    ));
