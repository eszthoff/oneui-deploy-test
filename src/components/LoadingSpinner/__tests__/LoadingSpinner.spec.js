import React from 'react';
import toJson from 'enzyme-to-json';
import LoadingSpinner from '../LoadingSpinner';

describe('<LoadingSpinner> that renders a circular loading spinner', () => {
    it('should render a default spinner correctly', () => {
        const wrapper = shallow(<LoadingSpinner />);
        expect(toJson(wrapper)).toMatchSnapshot();
    });

    it('should add classes when props are changed', () => {
        const wrapper = shallow(<LoadingSpinner context="primary" size={ 12 } />);
        expect(toJson(wrapper)).toMatchSnapshot();
    });
});
