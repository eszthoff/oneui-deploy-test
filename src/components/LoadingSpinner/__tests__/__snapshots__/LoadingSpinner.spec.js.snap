// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<LoadingSpinner> that renders a circular loading spinner should add classes when props are changed 1`] = `
<svg
  className="LoadingSpinner LoadingSpinner--context_primary LoadingSpinner--hidden"
  style={
    Object {
      "height": 12,
      "width": 12,
    }
  }
  viewBox={
    Array [
      0,
      0,
      44,
      44,
    ]
  }
>
  <circle
    className="LoadingSpinner__path LoadingSpinner__path--context_primary LoadingSpinner__path--hidden"
    cx="22"
    cy="22"
    fill="none"
    r="20"
    strokeWidth="4"
  />
</svg>
`;

exports[`<LoadingSpinner> that renders a circular loading spinner should render a default spinner correctly 1`] = `
<svg
  className="LoadingSpinner LoadingSpinner--context_brand"
  style={null}
  viewBox={
    Array [
      0,
      0,
      44,
      44,
    ]
  }
>
  <circle
    className="LoadingSpinner__path LoadingSpinner__path--context_brand"
    cx="22"
    cy="22"
    fill="none"
    r="20"
    strokeWidth="4"
  />
</svg>
`;
